File: android/app/src/main/kotlin/com/claudeos/api/Tools.kt
package com.claudeos.api

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.add
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonArray
import kotlinx.serialization.json.putJsonObject

enum class RiskTier { LOW, CONFIRM, SLIDE_TO_CONFIRM }

data class OsTool(
    val name: String,
    val description: String,
    val risk: RiskTier,
    val inputSchema: JsonObject,
)

object Tools {
    val all: List<OsTool> = listOf(
        tool(
            "render_ui",
            "Render a screen for the user. Call this at the end of every turn unless your reply is purely a brief spoken response. Replaces the current screen.",
            RiskTier.LOW,
            required = listOf("spec"),
        ) {
            putJsonObject("spec") {
                put("type", "object")
                put("description", "A UISpec: { speak?, title?, root, suggestions? }. Node types: text, stack, row, card, button, slide_to_confirm, list, image, photo_grid, field, form, toggle, slider, map, metric, chips, divider, spacer, media_player, camera_viewfinder, loading, error. Actions on buttons/slide_to_confirm: { kind: 'send', prompt } | { kind: 'tool', tool, input, risk } | { kind: 'open_url', url } | { kind: 'dismiss' }.")
            }
        },
        tool("remember", "Save a durable fact about the user (relationships, preferences, routines). Don't store secrets.",
            RiskTier.LOW, required = listOf("key", "value")) {
            putJsonObject("key") { put("type", "string"); put("description", "Stable short key like 'mom.phone' or 'coffee.order'") }
            putJsonObject("value") { put("type", "string") }
            putJsonObject("category") { put("type", "string"); put("description", "person | preference | routine | place | credential_ref | other") }
        },
        tool("recall", "Look up remembered facts by key prefix or category.", RiskTier.LOW) {
            putJsonObject("key_prefix") { put("type", "string") }
            putJsonObject("category") { put("type", "string") }
        },
        tool("add_lesson", "Save a short behavioral lesson for next time (e.g. 'when X, do Y'). Use after meaningful interactions.",
            RiskTier.LOW, required = listOf("lesson")) {
            putJsonObject("lesson") { put("type", "string") }
        },
        tool("send_message", "Send an SMS or chat message. Risky — invoke via slide_to_confirm.",
            RiskTier.SLIDE_TO_CONFIRM, required = listOf("recipient", "body")) {
            putJsonObject("recipient") { put("type", "string") }
            putJsonObject("body") { put("type", "string") }
            putJsonObject("channel") { put("type", "string"); put("description", "sms | whatsapp | imessage | telegram") }
        },
        tool("place_call", "Place a phone call. Risky — invoke via slide_to_confirm.",
            RiskTier.SLIDE_TO_CONFIRM, required = listOf("recipient")) {
            putJsonObject("recipient") { put("type", "string") }
        },
        tool("take_photo", "Open the camera viewfinder so the user can capture a photo.", RiskTier.LOW) {
            putJsonObject("reason") { put("type", "string"); put("description", "Why you need the photo") }
        },
        tool("search_photos", "Search the user's photo library by date, place or contents.",
            RiskTier.LOW, required = listOf("query")) {
            putJsonObject("query") { put("type", "string") }
        },
        tool("control_smart_home", "Control a smart home device. Reversible: confirm. Irreversible: slide_to_confirm.",
            RiskTier.CONFIRM, required = listOf("device", "command")) {
            putJsonObject("device") { put("type", "string") }
            putJsonObject("command") { put("type", "string"); put("description", "e.g. 'on' | 'off' | 'set:50%' | 'play:jazz'") }
        },
        tool("navigate", "Open turn-by-turn navigation.",
            RiskTier.LOW, required = listOf("destination")) {
            putJsonObject("destination") { put("type", "string") }
            putJsonObject("mode") { put("type", "string"); put("description", "driving | walking | transit | cycling") }
        },
        tool("pay", "Make a payment. Always slide_to_confirm.",
            RiskTier.SLIDE_TO_CONFIRM, required = listOf("recipient", "amount", "currency")) {
            putJsonObject("recipient") { put("type", "string") }
            putJsonObject("amount") { put("type", "number") }
            putJsonObject("currency") { put("type", "string") }
            putJsonObject("memo") { put("type", "string") }
        },
        tool("open_app", "Open an installed app by name.",
            RiskTier.LOW, required = listOf("app")) {
            putJsonObject("app") { put("type", "string") }
        },
        tool("calendar_read", "Read upcoming calendar events.", RiskTier.LOW) {
            putJsonObject("start") { put("type", "string") }
            putJsonObject("end") { put("type", "string") }
        },
        tool("calendar_write", "Create or modify a calendar event.",
            RiskTier.CONFIRM, required = listOf("title", "start")) {
            putJsonObject("title") { put("type", "string") }
            putJsonObject("start") { put("type", "string") }
            putJsonObject("end") { put("type", "string") }
            putJsonObject("location") { put("type", "string") }
            putJsonObject("notes") { put("type", "string") }
        },
        tool("web_browse", "Fetch a page or run a web search.",
            RiskTier.LOW, required = listOf("query_or_url")) {
            putJsonObject("query_or_url") { put("type", "string") }
        },
        tool("read_screen", "Read the current foreground app's on-screen content (text + clickable elements). Requires the user to have enabled the Claude OS accessibility service. Useful for inspecting WhatsApp, Maps, banking, or any other app the user is currently viewing.",
            RiskTier.LOW) {
        },
        tool("read_recent_notifications", "Return recent notifications captured by the launcher's notification listener (texts, WhatsApp/Slack/etc messages, calendar reminders). Requires the user to have enabled notification access for Claude OS.",
            RiskTier.LOW) {
            putJsonObject("limit") { put("type", "integer"); put("description", "Max notifications to return (default 20)") }
            putJsonObject("from_app") { put("type", "string"); put("description", "Optional package-name substring filter, e.g. 'whatsapp', 'gmail', 'messaging'") }
        },
        tool("wait_ms", "Pause for a given number of milliseconds. Useful between open_app and read_screen so the launched app has time to render.",
            RiskTier.LOW, required = listOf("ms")) {
            putJsonObject("ms") { put("type", "integer"); put("description", "Milliseconds to wait, capped at 5000.") }
        },
        tool("interrupt_user", "Show a prominent in-launcher notification interrupting the user with an important message (incoming text from a key person, calendar reminder, etc).",
            RiskTier.LOW, required = listOf("headline")) {
            putJsonObject("headline") { put("type", "string") }
            putJsonObject("body") { put("type", "string") }
            putJsonObject("from") { put("type", "string") }
            putJsonObject("urgency") { put("type", "string"); put("description", "low | normal | high") }
        },
    )

    val byName: Map<String, OsTool> = all.associateBy { it.name }
    val serverHandled: Set<String> = setOf("render_ui", "remember", "recall", "add_lesson")

    fun anthropicSchema(): JsonArray = buildJsonArray {
        all.forEach { t ->
            add(buildJsonObject {
                put("name", t.name)
                put("description", t.description)
                put("input_schema", t.inputSchema)
            })
        }
    }
}

private fun tool(
    name: String,
    description: String,
    risk: RiskTier,
    required: List<String> = emptyList(),
    properties: kotlinx.serialization.json.JsonObjectBuilder.() -> Unit,
): OsTool {
    val schema = buildJsonObject {
        put("type", "object")
        putJsonObject("properties") { properties() }
        if (required.isNotEmpty()) {
            putJsonArray("required") { required.forEach { add(JsonPrimitive(it)) } }
        }
    }
    return OsTool(name, description, risk, schema)
}

File: src/lib/os/tools.ts
// Tool palette Claude can call.
// Every tool has a risk tier — `slide_to_confirm` tools must be invoked
// through a slide_to_confirm UI node, not directly.

import type { Tool } from "@anthropic-ai/sdk/resources/messages";

export type RiskTier = "low" | "confirm" | "slide_to_confirm";

export interface OsToolDef {
  name: string;
  description: string;
  risk: RiskTier;
  input_schema: Tool["input_schema"];
}

// The renderer is the one that "talks" to the user — Claude almost always
// finishes a turn by calling `render_ui`. Other tools perform real-world side
// effects on the device.
export const TOOLS: OsToolDef[] = [
  {
    name: "render_ui",
    description:
      "Render a screen for the user. Call this at the end of every turn unless you only need to speak. The UI you return replaces the current screen. Be concise and visually clean — pick the smallest set of components that conveys the answer or asks the user for what's needed next.",
    risk: "low",
    input_schema: {
      type: "object",
      properties: {
        spec: {
          type: "object",
          description:
            "A UISpec. Shape: { speak?: string, title?: string, root: Node, suggestions?: string[] }. Node is one of: text, stack, row, card, button, slide_to_confirm, list, image, photo_grid, field, form, toggle, slider, map, metric, chips, divider, spacer, media_player, camera_viewfinder, loading, error. Buttons and slide_to_confirm carry an `action`: { kind: 'send', prompt } | { kind: 'tool', tool, input, risk, label } | { kind: 'open_url', url } | { kind: 'dismiss' }.",
        },
      },
      required: ["spec"],
    },
  },
  {
    name: "remember",
    description:
      "Save a fact about the user to long-term personal memory. Use sparingly — only durable facts (relationships, preferences, routines, IDs). Do NOT store sensitive secrets.",
    risk: "low",
    input_schema: {
      type: "object",
      properties: {
        key: { type: "string", description: "Short stable key, e.g. 'mom.phone' or 'coffee.order'" },
        value: { type: "string", description: "The value to remember" },
        category: {
          type: "string",
          enum: ["person", "preference", "routine", "place", "credential_ref", "other"],
        },
      },
      required: ["key", "value"],
    },
  },
  {
    name: "recall",
    description: "Look up a remembered fact by key prefix or category.",
    risk: "low",
    input_schema: {
      type: "object",
      properties: {
        key_prefix: { type: "string" },
        category: { type: "string" },
      },
    },
  },
  {
    name: "send_message",
    description: "Send an SMS, iMessage, or chat-app message. Risky — must be invoked via slide_to_confirm.",
    risk: "slide_to_confirm",
    input_schema: {
      type: "object",
      properties: {
        recipient: { type: "string" },
        body: { type: "string" },
        channel: { type: "string", enum: ["sms", "whatsapp", "imessage", "telegram"] },
      },
      required: ["recipient", "body"],
    },
  },
  {
    name: "place_call",
    description: "Place a phone call. Risky — must be invoked via slide_to_confirm.",
    risk: "slide_to_confirm",
    input_schema: {
      type: "object",
      properties: { recipient: { type: "string" } },
      required: ["recipient"],
    },
  },
  {
    name: "take_photo",
    description: "Open the camera viewfinder so the user can capture a photo. Renders a camera UI; the user taps to capture.",
    risk: "low",
    input_schema: {
      type: "object",
      properties: { reason: { type: "string", description: "Why you need the photo" } },
    },
  },
  {
    name: "search_photos",
    description: "Search the user's photo library by date, place or contents (e.g. 'Italy', 'last weekend').",
    risk: "low",
    input_schema: {
      type: "object",
      properties: { query: { type: "string" } },
      required: ["query"],
    },
  },
  {
    name: "control_smart_home",
    description: "Control a smart home device. Reversible actions are 'confirm', irreversible (locks, garage) are slide_to_confirm.",
    risk: "confirm",
    input_schema: {
      type: "object",
      properties: {
        device: { type: "string" },
        command: { type: "string", description: "e.g. 'on', 'off', 'set:50%', 'play:jazz'" },
      },
      required: ["device", "command"],
    },
  },
  {
    name: "navigate",
    description: "Open turn-by-turn navigation to a destination.",
    risk: "low",
    input_schema: {
      type: "object",
      properties: {
        destination: { type: "string" },
        mode: { type: "string", enum: ["driving", "walking", "transit", "cycling"] },
      },
      required: ["destination"],
    },
  },
  {
    name: "pay",
    description: "Make a payment. ALWAYS slide_to_confirm.",
    risk: "slide_to_confirm",
    input_schema: {
      type: "object",
      properties: {
        recipient: { type: "string" },
        amount: { type: "number" },
        currency: { type: "string" },
        memo: { type: "string" },
      },
      required: ["recipient", "amount", "currency"],
    },
  },
  {
    name: "open_app",
    description: "Open a phone app by name (Spotify, Maps, Calendar, etc).",
    risk: "low",
    input_schema: {
      type: "object",
      properties: { app: { type: "string" } },
      required: ["app"],
    },
  },
  {
    name: "calendar_read",
    description: "Read upcoming calendar events.",
    risk: "low",
    input_schema: {
      type: "object",
      properties: {
        start: { type: "string", description: "ISO date" },
        end: { type: "string", description: "ISO date" },
      },
    },
  },
  {
    name: "calendar_write",
    description: "Create or modify a calendar event. Confirm tier.",
    risk: "confirm",
    input_schema: {
      type: "object",
      properties: {
        title: { type: "string" },
        start: { type: "string" },
        end: { type: "string" },
        location: { type: "string" },
        notes: { type: "string" },
      },
      required: ["title", "start"],
    },
  },
  {
    name: "add_lesson",
    description:
      "Save a short behavioral lesson to long-term memory so you do better next time. Examples: 'when the user asks about photos, default to a 3-column grid', 'mom prefers WhatsApp over SMS', 'after dim-the-lights, also ask if they want music'. Use after every meaningful interaction where you noticed something to improve.",
    risk: "low",
    input_schema: {
      type: "object",
      properties: {
        lesson: { type: "string", description: "The lesson, written as a short imperative." },
      },
      required: ["lesson"],
    },
  },
  {
    name: "propose_code_change",
    description:
      "Propose a patch to your own source code to add a new capability, fix a bug, or improve a UI component. Returns a proposal the user must approve via slide-to-confirm. Only works when running in the dev environment. Use this when you hit a limitation you can't work around — e.g. a missing UI component, a bug in the renderer, a tool you wish you had. Provide one or more file edits as old_string/new_string pairs (exact match) or full file rewrites. Be precise — old_string must match exactly.",
    risk: "slide_to_confirm",
    input_schema: {
      type: "object",
      properties: {
        rationale: { type: "string", description: "Why this change is needed, in 1-2 sentences for the user." },
        edits: {
          type: "array",
          description: "List of edits to apply atomically.",
          items: {
            type: "object",
            properties: {
              path: { type: "string", description: "Repo-relative path, e.g. 'src/components/os/Renderer.tsx'." },
              mode: { type: "string", enum: ["edit", "create", "rewrite"], description: "edit = swap old_string for new_string; create = new file; rewrite = replace whole file." },
              old_string: { type: "string", description: "Exact substring to find (edit only)." },
              new_string: { type: "string", description: "Replacement (edit), or full file content (create/rewrite)." },
            },
            required: ["path", "mode", "new_string"],
          },
        },
      },
      required: ["rationale", "edits"],
    },
  },
  {
    name: "web_browse",
    description: "Fetch and summarize a web page or run a web search.",
    risk: "low",
    input_schema: {
      type: "object",
      properties: { query_or_url: { type: "string" } },
      required: ["query_or_url"],
    },
  },
];

export function anthropicTools() {
  return TOOLS.map((t) => ({
    name: t.name,
    description: t.description,
    input_schema: t.input_schema,
  }));
}

export function toolByName(name: string): OsToolDef | undefined {
  return TOOLS.find((t) => t.name === name);
}
